feat: energy consumption and instantaneous power sensors

christian-nils opened the pull request:
This PR implements the instantaneous power, and the energy consumption sensors.
This is the first time I am writing a sensor for a HA integration, let me know if there is anything I should improve.

Thanks

/CN

toringer commented:
Thanks 👍 Will have a look at it during the weekend

toringer commented on custom_components/heru/sensor.py:
```diff
@@ -103,11 +109,122 @@ async def async_setup_entry(
+                self._exhaust_p_entity,
+            ]
+        ):
+            heater_p = float(self.hass.states.get(self._heater_p_entity).state)
```
Would it be better to subscribe to state changes and disable scan_interval for the new sensors? See https://developers.home-assistant.io/docs/integration_listen_events/ 

Since all sensors are polling every 15 seconds, including the power and energy sensor, it might use "old" values.

christian-nils replied:
I agree, the power sensor could be event-based. But I am not sure this could applicable to the energy consumption, if the power is constant over a long period of time you will not see any increase in the energy consumption.
I will implement the event-based trigger for the power, what do you think for the energy sensor?

toringer replied:
I was thinking of a scenario where all sensors for the HERU component would start polling at the same time. Since the modbus sensor most likely will take longer time to get new values because of network time, the power and energy sensor would finish executing before new values are retrieved from the HERU unit, and thereby using values from the previous poll. 

Having the energy sensor event-based might mess up the calculation as the time is no longer constant? My electro skills are not that good by the way :-)

Lets start with the power sensor for now 👍 Looking forward to testing it!

christian-nils replied:
I have implemented the event-based update for the power sensor. I kept the energy consumption sensor's data polling. We would have to track the time if we change it and the data of the energy sensor needs to be updated periodically anyway for the energy dashboard to have the right statistics.
Please test and let me know if there is anything to improve! Thanks, have a great weekend.

toringer commented on custom_components/heru/config_flow.py:
```diff
@@ -66,8 +69,18 @@ async def _show_config_form_user(self, user_input):
 
         user_schema = {
             vol.Required(CONF_DEVICE_NAME): cv.string,
+            vol.Required(
```
This should be added to the OptionsFlowHandler method as well to support changing the model.

Same for the port data type bellow

christian-nils replied:
Ok, I have done it. Is there some sort of reloading done automatically to read the correct unit model after the change? 

For instance, the modbus client seems to be created only in `async_setup_entry`, but this seems to be called only at the integration initialization.

christian-nils replied:
I do not think it is 100% flawless, but I have implemented it and it worked in different use cases. You will need to restart HASS to account for the change of the model.

toringer replied:
> Ok, I have done it. Is there some sort of reloading done automatically to read the correct unit model after the change?
> 
> For instance, the modbus client seems to be created only in `async_setup_entry`, but this seems to be called only at the integration initialization.

Hm, not sure actually. There are some unused methods in __init__.py: ```async_unload_entry```` and ```async_reload_entry``` (from the code base I based the HERU code on...). Could it be that the ```OptionsFlowHandler``` should do its own recreate?

toringer replied:
I will look into the issues when saving after reconfiguring in #9

toringer commented on custom_components/heru/config_flow.py:
The VERSION needs to be bumped. Otherwise, users upgrading from a previous version will get an error when the component load as the unit model would be None. See https://developers.home-assistant.io/docs/config_entries_config_flow_handler#config-entry-migration

christian-nils replied:
Good point, I will have a look.

christian-nils replied:
Alright, I did something.

toringer commented:
Getting this error when adding the HERU device (removed the device and re-added it)
```
Traceback (most recent call last):
  File "/home/vscode/.local/lib/python3.10/site-packages/homeassistant/helpers/entity.py", line 541, in async_update_ha_state
    await self.async_device_update()
  File "/home/vscode/.local/lib/python3.10/site-packages/homeassistant/helpers/entity.py", line 739, in async_device_update
    raise exc
  File "/workspaces/home-assistant-heru/.devcontainer/hass/custom_components/heru/sensor.py", line 232, in async_update
    float(self.hass.states.get(self._power_entity).state)
TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'
```

```Instantaneous power``` has a value, but ```Energy consumption``` is "Unknown". After a restart of Home Assistant, the error is gone and ```Energy consumption``` gets a value.

toringer commented:
By the way. I plugged the unit into a Datek APEX smart plug 16A which report wattage. It shows 11W while the ```Instantaneous power``` sensor shows 122W... I have a HERU 100

christian-nils commented:
> By the way. I plugged the unit into a Datek APEX smart plug 16A which report wattage. It shows 11W while the `Instantaneous power` sensor shows 122W... I have a HERU 100

That is strange. What is your supply and exhaust fan power (in %)?  I don't have a power measurement device but I do have an energy meter from my electricity provider, I checked the instantaneous power from the phase on which my Heru 100S is plugged in and I have the following when activating/deactivating the boost mode:

![image](https://github.com/toringer/home-assistant-heru/assets/2658014/865b4b57-c6b1-4caa-b22b-c84017303af2)

My energy consumption equation estimate an increase of 47W when activating the boost, the increase on the energy meter is around 50-60W (the precision of the value is 0.2f in kW). So I think the equation is not too off. I will buy some Shelly Plus 1PM so I can have a more accurate measurement on the unit. 

At the end of the day, if we feel that the feature does not bring any improvement (because too inaccurate), we can definitively scrap it :-)

christian-nils commented:
> Getting this error when adding the HERU device (removed the device and re-added it)
> 
> ```
> Traceback (most recent call last):
>   File "/home/vscode/.local/lib/python3.10/site-packages/homeassistant/helpers/entity.py", line 541, in async_update_ha_state
>     await self.async_device_update()
>   File "/home/vscode/.local/lib/python3.10/site-packages/homeassistant/helpers/entity.py", line 739, in async_device_update
>     raise exc
>   File "/workspaces/home-assistant-heru/.devcontainer/hass/custom_components/heru/sensor.py", line 232, in async_update
>     float(self.hass.states.get(self._power_entity).state)
> TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'
> ```
> 
> `Instantaneous power` has a value, but `Energy consumption` is "Unknown". After a restart of Home Assistant, the error is gone and `Energy consumption` gets a value.
> 
> The same error occurs if I remove and re-add again, so you should be able to reproduce the error

Ok, should be fixed with https://github.com/toringer/home-assistant-heru/pull/7/commits/63281eb8d9c40a34a1cd011a95e43bea78826e86

toringer commented:
> > By the way. I plugged the unit into a Datek APEX smart plug 16A which report wattage. It shows 11W while the `Instantaneous power` sensor shows 122W... I have a HERU 100
> 
> That is strange. What is your supply and exhaust fan power (in %)? I don't have a power measurement device but I do have an energy meter from my electricity provider, I checked the instantaneous power from the phase on which my Heru 100S is plugged in and I have the following when activating/deactivating the boost mode:
> 
> ![image](https://user-images.githubusercontent.com/2658014/239764296-865b4b57-c6b1-4caa-b22b-c84017303af2.png)
> 
> My energy consumption equation estimate an increase of 47W when activating the boost, the increase on the energy meter is around 50-60W (the precision of the value is 0.2f in kW). So I think the equation is not too off. I will buy some Shelly Plus 1PM so I can have a more accurate measurement on the unit.
> 
> At the end of the day, if we feel that the feature does not bring any improvement (because too inaccurate), we can definitively scrap it :-)

I just realised that my smart plug probably will not measure inductive load. So that might be the reason for the low wattage.

toringer commented:
I'm ready to merge 👍  But could you first add the french translation for ```device_model``` and possible add a note in the README that the two sensors are calculated?

I will add a note in the release note that the component needs to be reconfigured for existing installs to enable the sensors.

christian-nils commented:
> I'm ready to merge +1 But could you first add the french translation for `device_model` and possible add a note in the README that the two sensors are calculated?
> 
> I will add a note in the release note that the component needs to be reconfigured for existing installs to enable the sensors.

I have added the FR translation and updated the README.md file. Before merging the PR, I would like to compare the energy consumption we calculate with what I will measure with the Shelly Plus 1 PM. I will receive them tomorrow, I can do the test over the weekend.

christian-nils commented:
I actually feel a bit embarassed of having been so naive, but it is not that straightforward to estimate the energy consumption as it depends on some factors that we do not have information about, for instance the static pressure or the airflow. If you look at https://fsp.ostberg.com/products/4396/heru-100-s-ec/?region=se&lang=en under the **Diagrams** section, you can see that given the static pressure, the airflow, and the voltage you will get the current power, which is very far from the equation I have implemented in my PR.

I suggest to reject the PR as we miss the needed variables to have a good energy consumption estimate. Sorry for the troubles!
